### PR TITLE
Add autodiscovery and port selection features, added number and climate entities, added some tests (examples)

### PR DESCRIPTION
**Pull request by mjcumming**
Hi Peter,

Thank you for starting this project. I have projects that are PI based that I am going to move to this library from MQTT. I have been using this project with OpenHAB and the ESPHome binding. Additions I made include adding the ability to specify the ports for the device and webserver, autodiscovery, light and climate entities, and some test files.

Mike

### CODE REVIEW

File: aioesphomeserver/device.py [hunk 1]
```diff
@@ -20,29 +23,48 @@ def __init__(
...
             friendly_name=None,
-            suggested_area=None
+            suggested_area=None,
+            network=None,
+            board=None,
+            platform=None
     ):
         self.name = name
-        self.mac_address = mac_address or self._generate_mac_address
+        self.mac_address = mac_address or self._generate_mac_address()
```
Comment (peterkeen) [inline aioesphomeserver/device.py:32]: Good catch

File: aioesphomeserver/device.py [hunk 2]
```diff
@@ -79,13 +101,56 @@ def get_entity_by_key(self, key):
...
+        self.add_entity(WebServer(name="_web_server", port=self.web_port))
+
+        self.zeroconf = await self.register_zeroconf(self.api_port)
 
         async with asyncio.TaskGroup() as tg:
             for entity in self.entities:
                 if hasattr(entity, 'run'):
                     tg.create_task(entity.run())
+
+    async def register_zeroconf(self, port):
```
Comment (peterkeen) [inline aioesphomeserver/device.py:120]: Could you talk a bit about why this is both here and in BasicEntity? It looks largely the same except for a couple differences in the properties.

**Review by peterkeen — COMMENTED** [review]
Thanks so much! Just have a question about the zero conf stuff. 

--Pete

File: aioesphomeserver/device.py (hunk 2 above)
Comment (mjcumming) [reply aioesphomeserver/device.py:120]: Agree, it's duplicated. I'll move all the zeroconf code to the device file and remove from the basic entity.

**Comment by mjcumming** [comment]
I'll make the changes and push.

**Pull request closed without merging**